### DISCUSSION TITLE
UPnP implementation

### DISCUSSION THREAD

**Discussion by StefanoSetti** [libp2p/rust-libp2p#2331] [Q&A] [answered]
is there any implementation of UPnP protocol in this library?

**Comment by mxinden** [comment, marked as answer]
Not yet. Contributions would be very much appreciated.

A couple of pointers:

- Past discussion on UPnP https://github.com/libp2p/rust-libp2p/issues/558
- General NAT traversal tracking issue for rust-libp2p including UPnP https://github.com/libp2p/rust-libp2p/issues/2052
- Support in go-libp2p https://github.com/libp2p/go-libp2p/blob/e70d7cf245fd80b80fb539bd6753e9a8c826cb6f/options.go#L347-L351

**Reply by StefanoSetti** [reply]
Thanks for answering. So, right now is not possible for me to implement it, I'll do it if I will have the chance. In any case, is [this pull](https://github.com/libp2p/rust-libp2p/pull/2262) in any way related?

**Reply by mxinden** [reply]
Related, as it is one piece in the greater hole punching effort, though it will not enable you to do UPnP.
